Advanced Walkthroughs (#484)
permitio/docs: docs/overview/create-a-rebac-policy.mdx

---
sidebar_position: 6
title: Create a ReBAC Policy
---

import WhatsNext from "../../src/components/whats-next/WhatsNext";
import TimelineStep from "../../src/components/timeline/TimelineStep";
import TimelineWrapper from "../../src/components/timeline/TimelineWrapper";
import Video from "../../src/components/video/Video";

Setting up Relationship-Based Access Control (ReBAC) policies may seem complex, but with Permit, it’s a simple and intuitive
process. This step-by-step guide will walk you through configuring and enforcing ReBAC policies using the Permit dashboard.

:::info Before You Begin

> This guide assumes you have already set up your Permit account and have a foundational understanding of ReBAC and the basics

Reviewer: Please add links to read more about the basic understanding of ReBAC

of how Permit operates.
:::

## Defining our ReBAC Policy

In this guide, we'll set up ReBAC permissions for a dashboarding system with two resource types: `Dashboards (parent)` and
`Widgets (child)`. Users can have roles — `Owner`, `Viewer`, or `Analyst/Editor` — each with specific permissions.

The access to a **Dashboard resource** will automatically includes permissions for its **associated Widgets**.

> ### Visual representation

Reviewer: Redundant title

```suggestion
```

<br />

> ```mermaid

Reviewer: This diagram still shows a mix of the schema and the data, please decouple

    Widget2[Widget 2]:::resource
    WidgetN[Widget N]:::resource

    %% Roles
    Owner[Owner]:::role
    Viewer[Viewer]:::role
    AnalystEditor[Analyst/Editor]:::role

    %% Resource Hierarchy
    Dashboard --> Widget1
    Dashboard --> Widget2
    Dashboard --> WidgetN

    %% Explicit Permissions for Dashboard
    Dashboard -->|Owner| Owner
    Dashboard -->|Viewer| Viewer
    Dashboard -->|Analyst/Editor| AnalystEditor

    %% Explicit Permissions for Widgets
    Widget1 -->|Owner| Owner
    Widget1 -->|Viewer| Viewer
    Widget1 -->|Analyst/Editor| AnalystEditor

    Widget2 -->|Owner| Owner
    Widget2 -->|Viewer| Viewer
    Widget2 -->|Analyst/Editor| AnalystEditor

    WidgetN -->|Owner| Owner
    WidgetN -->|Viewer| Viewer
    WidgetN -->|Analyst/Editor| AnalystEditor

    %% Implicit Permissions (Inheritance)
    Dashboard -.->|Inherits Roles| Widget1
    Dashboard -.->|Inherits Roles| Widget2
    Dashboard -.->|Inherits Roles| WidgetN

    %% Link Styles
    linkStyle 4,5,6 stroke:#333, stroke-width:2px

    %% Implicit Permissions (Inheritance)
    linkStyle 7,8,9 stroke:#0f0, stroke-width:2px, stroke-dasharray: 5,5
```

<br />
<br />

> ## Step 1: Create Resources and their Roles

Reviewer: I'm missing here 2 contexts:
1. The following steps will configure the ReBAC policy stems/schema
2. The `Step 1:` declared here, should have context of the journey that `Step 1:` is part of

Reviewer: I think if we will replace the order of the first paragraph (so first, `For this example` and then `Start by navigating`

---

Start by navigating to the Resources section in the Permit dashboard - `Policy > Resources`. Once there, locate the `Add Resource`
button to begin creating your first resource. For this example, we’ll define two resources: `Dashboard` and `Widget`, each with
specific actions and resource roles.

> <img

Reviewer: I think this image is redundant as you show it in the next video

<img src="/ui-videos/walkthroughs/rebac/resources.png" alt="Dashboard Resource" data-zoomable />

</TimelineStep>
</TimelineWrapper>

## Step 2: Define resource relations

The next step it to configure the relations between resources within the Permit Dashboard.

---

<TimelineWrapper>
<TimelineStep>

### Open the Resource Editor

In the Resources section, locate the `Dashboard` resource. Click the `three-dot menu icon` next to it and select `Edit`.

<img src="/ui-videos/walkthroughs/rebac/edit-resource.png" alt="Edit Resource" data-zoomable />

</TimelineStep>
<TimelineStep>

### Define the Parent-Child Relation

Navigate to the `Relations` section of the resource editor. Configure the `Dashboard resource as the parent of the Widget resource`.
This setup ensures that the Dashboard can act as a container for its associated Widget resources.

<img
  src="/ui-videos/walkthroughs/rebac/resource-relations.png"
  alt="Resource Relations"
  data-zoomable
/>

:::note
A resource cannot be its own parent. If a resource requires a self-relation, consider using an alternative relation type such as
owner or container.
:::

</TimelineStep>
<TimelineStep>

### Save the Relation

> Once the relationship is defined, save the configuration. You now have a clear hierarchical relation where Dashboard

Reviewer: ```suggestion
Once a relation is defined, save the configuration. You now have a clear hierarchical relation where the Dashboard
```

> resources are parents to Widget resources.

Reviewer: ```suggestion
resource are parents to Widget resource.
```

> :::info

Reviewer: I think this admonition is belongs to the previous section

Once you successfully define the relation on the roles - it will automatically create the roles for you in the policy
editor.
:::

</TimelineStep>
</TimelineWrapper>

## Step 3: Define role derivations

> In this step, we will be configuring the role derivations that will exist between the various user roles and their associated

Reviewer: Please add a link to a doc that explains what role derivation is

relationships within our system

---

<TimelineWrapper>
<TimelineStep>

### Navigate to the Resource’s Role Derivation Settings

In the Permit dashboard, go to `Policy > Roles` where you should now see all `Resource Roles`.

<img src="/ui-videos/walkthroughs/rebac/roles-screen.png" alt="Roles Screen" data-zoomable />

</TimelineStep>
<TimelineStep>

### Define Role Derivation Logic

Within the Role Derivation section, establish the rules for how roles are inherited or assigned:

- For the Widget resource, set rules that derive roles from its parent Dashboard resource.
  Example configurations:

>   - If a user is an `Owner` of a `Dashboard`, they automatically become an `Owner` of `all Widget instances` under that Dashboard.

Reviewer: ```suggestion
  - If a user is assigned as an `Owner` of a `Dashboard` instance, they automatically become an `Edutir` of `all Widget` under that Dashboard.
```

>     <Video

Reviewer: Most of the steps in this video are redundant, as when you assign a user as an owner/editor to the widget, they do not need to be a viewer.
Please fix it according to our discussion

      src={"/ui-videos/walkthroughs/rebac/role-derivation-logic-rebac.mp4"}
      controls
      autoPlay={false}
    />

>   - If a user is a `Viewer` of a `Dashboard`, they inherit the `Viewer` role for `all its Widget instances`.

Reviewer: See previous comment about the terminology of resource and instance

    <img
      src="/ui-videos/walkthroughs/rebac/dashboard-viewer-derivations.png"
      alt="Dashboard Viewer Derivations"
      data-zoomable
    />

</TimelineStep>
<TimelineStep>

### Save the Role Derivation Rules

Once the derivation logic is defined, save the changes to apply the configuration.

> <img

Reviewer: As the video comment, it does not need all the derivations

  data-zoomable
/>

<img
  src="/ui-videos/walkthroughs/rebac/dashboard-viewer-derivation-2.png"
  alt="Dashboard Viewer Derivation Summary"
  data-zoomable
/>

</TimelineStep>
</TimelineWrapper>

## Step 4: Configure permissions

For each role, assign the actions that the role can perform on the particular resource resource.
For this, navigate to `Policy`.

---

<TimelineWrapper>
<TimelineStep>

### For the Dashboard resource:

- **Owner**: Can `view`, `create`, `edit`, `delete`, `add-widget`, and `remove-widget`.
- **Viewer**: Can `view` only.
- **Analyst**: Can `view`, `add-widget`, and `edit`.

  <img
    src="/ui-videos/walkthroughs/rebac/dashboard-permissions.png"
    alt="Dashboard Permissions"
    data-zoomable
  />

</TimelineStep>
<TimelineStep>

### For the Widget resource:

- **Owner**: Can `view`, `create`, `edit`, and `delete`.
- **Viewer**: Can only `view`.
- **Editor**: Can `view` and `edit`.

  <img
    src="/ui-videos/walkthroughs/rebac/widget-permissions.png"
    alt="Widget Permissions"
    data-zoomable
  />

Once you assigned and saved your permissions, it's now time to test our policy and make sure it works.

</TimelineStep>
</TimelineWrapper>

> ## Step 5: Testing our Policy

Reviewer: This should be a new step 1 i think, because here we start sync data. 
Also, here, as previously, it is missing the context of `we finish the schema, let's start with data.`

In this step, we will create users

> ```mermaid

Reviewer: The mermaid is good, but the layout is hard to understand. I recommend go with excalidraw and put the users on top.

    WidgetA[Widget A]:::resource
    WidgetB[Widget B]:::resource

    %% Relationships
    John -->|Owner| Dashboard1
    Anna -->|Viewer| Dashboard1

    Dashboard1 --> WidgetA
    Dashboard1 --> WidgetB

    WidgetA -->|Analyst/Editor| Anna
    WidgetB -->|Viewer| John
```

---

<TimelineWrapper>
<TimelineStep>

### Create Users

We need access to test users to validate our policy and ensure a specific user complies with the defined rules.

Let's create two users:

1. `John Smith` - john@smith.com
2. `Anna Smith` - anna@smith.com

They will have **no** `top-level` roles. We’ll **revisit** the instance access section during the testing phase.

<Video src={"/ui-videos/walkthroughs/rebac/creating-users-rebac.mp4"} controls autoPlay={false} />

</TimelineStep>
<TimelineStep>

### Create Resource Instances and their Relationship-Tuples

In this step, we will create an instance of the `Widget` resource called `Data Consumption` on a specific instance of `Dashboard`
resource called `Data`.

Navigate to `Directory`, click on `Instances` and click `Add Resource Instance`.

1. Select the **resource type** as `Dashboard`.
2. Pick your **instance key** unique identifier.

> 3. Select the `tenant` you would like to work within

Reviewer: Here it should focus on `default` as we do not explain what is `you would like to work within`

4. Create your relationship. Select the `type` from the dropdown list, and then fill in the `subject` and `object` instance.

:::tip OUR CONFIG
**Dashboard** `Data` is parent of **Widget** `Data Consumption`
:::

<Video
  src={"/ui-videos/walkthroughs/rebac/create-resource-instances-rebac.mp4"}
  controls
  autoPlay={false}
/>

</TimelineStep>
<TimelineStep>

### Assign Roles

Let's assign an `instance role` to `John Smith`. We want John to be the `owner` of the `data_consumption` widget.

- Navigate to `Directory` and `Add Instance Roles`.
- Scroll down, and give the user permissions per resource instance.

:::tip In our Cases
**`Widget`**:**`data_consumption`**#**`Owner`**
:::

<Video
  src={"/ui-videos/walkthroughs/rebac/assign-instance-roles-rebac.mp4"}
  controls
  autoPlay={false}
/>

</TimelineStep>
<TimelineStep>

> ### Test the Policy

Reviewer: As the comment on the step called `Test the policy` the titles here need some edit

We have now configured the whole ReBAC policy - and we have one last step left - which is testing if everything that we
have configured works.

> ```js

Reviewer: I recommend to break lines between the arguments so we can have a clear view of the check function
